### ISSUE TITLE
use color to differentiate results output

### ISSUE THREAD

**Issue by williballenthin** [mandiant/flare-floss#623]
we currently use a lot of ASCII art to break up the output in to sections and metadata. it can be a little overwhelming, especially when a user wants to find a string and work backwards. perhaps we can use a bit of color judiciuosly to highlight the interesting data and mute the rest:

<img width="452" alt="image" src="https://user-images.githubusercontent.com/156560/215755944-b3a37462-8fa0-48b1-b9e3-fa576cf0f0af.png">

**Comment by Arker123** [comment]
Hi @williballenthin , can I work on this issue?

**Comment by mr-tz** [comment]
Absolutely, I'll assign you. Thanks!

**Comment by Arker123** [comment]
Hi @mr-tz  I'm wondering if you could provide some feedback on the colors. I've highlighted the strings. I've included the screenshots of the output, and I would appreciate any suggestions you might have.
<p float="left">
<img width="50%" src="https://user-images.githubusercontent.com/94680887/223024650-db6bc9fc-4692-4576-9533-44ecfe7c4318.png">
<img width="44%" src="https://user-images.githubusercontent.com/94680887/223024730-f41d96f5-bb4f-47d6-ad02-a7999af5939f.png">
</p>

**Comment by mr-tz** [comment]
Cool, I'm all for colors, but would like to
* use them sparely, to focus just on the key parts, e.g., strings and maybe headings (see https://github.com/mandiant/flare-floss/pull/646#issuecomment-1455752036)
* have a solution that works on all configurations, especially light and dark backgrounds (maybe via a library?)

**Comment by Arker123** [comment]
Thank you for the feedback and suggestion @mr-tz !

I've made the changes you suggested and the output now uses color sparingly to highlight only the key parts such as strings and headings. Additionally, I've tested the solution on both light and dark backgrounds to ensure that it works on all configurations.

I have some designs to select from:-
1)  Green Strings with Colored Heading and muted subheadings
<p float="left">
<img width="50%" src="https://user-images.githubusercontent.com/94680887/223335682-ef626ba5-0229-4e82-905a-97ee38b9f946.png">
<img width="49%" src="https://user-images.githubusercontent.com/94680887/223335698-9d1709f5-a356-49c1-bacd-168c692a30fc.png">
</p>


2) Blue Strings with Colored Heading and muted subheadings
<p float="left">
<img width="50%" src="https://user-images.githubusercontent.com/94680887/223335777-40c9a198-82ec-4b35-85bb-cab87dfb8c54.png">
<img width="49%" src="https://user-images.githubusercontent.com/94680887/223335788-5ec848eb-79ae-4b71-8dac-38190415357c.png">
</p>


3) Yellow Strings with highlighted key part
<p float="left">
<img width="50%" src="https://user-images.githubusercontent.com/94680887/223335864-9c1d1fbe-e28b-4f4d-8615-4605d1b786aa.png">
<img width="49%" src="https://user-images.githubusercontent.com/94680887/223335852-a87b983b-6624-4bdc-b779-ce1c3303d35d.png">
</p>


4) Green Strings with highlighted key part
<p float="left">
<img width="50%" src="https://user-images.githubusercontent.com/94680887/223335904-23147b1d-11ed-4943-bb73-7a81b4c2fcac.png">
<img width="49%" src="https://user-images.githubusercontent.com/94680887/223335911-2e83bda4-8e3a-48f6-9921-0ff14eaebbc6.png">
</p>


Could you please let me know which designs to select from? I have these options and would appreciate your input in choosing the best one for this project. Thank you.

**Comment by mr-tz** [comment]
Wow, very cool. Thanks for the suggestions and screenshots. My preference would be: 1. (with no muting of the subheadings maybe as it looks rather faint?), 4., 2., and 3. but they all look good. Curious for your and other's opinions here!

For this I think we may want to add a command line argument `--color` to enable/disable these, see e.g. https://github.com/mandiant/capa/blob/master/capa/main.py#L813.

**Comment by williballenthin** [comment]
Maybe we can use the convention described here for tools like capa and FLOSS:
https://no-color.org/

**Comment by Shackles404** [comment]
I personally like the 4th one, because I think subheadings are probably too important to be muted and highlighting the important part (ASCII,..) looks good... Amazing colour choices though!

**Comment by williballenthin** [comment]
I also prefer the sub-headings to *not* be muted.

I have a minor preference for highlighting the key part of the titles, but its totally unscientific.

I sort of like the blues in the second one, but again for no real concrete reason.



All in all, though, great work experimenting with the palettes and sharing the screenshots!

**Comment by mr-tz** [comment]
Hey, that's great insight and I like the new output designs a lot!
One minor concern I have is whether the visual style would hinder some workflows (e.g. copy pasting the results somewhere else). 

However, we have the `-q` mode so could refer to that if analysts just want the data? Or should we add a new rendering argument to enable/disable this format? Curious to hear other thoughts here as well.

**Comment by williballenthin** [comment]
> Would it be okay if I go ahead and make a pull request with the changes I've made, along with the screenshots?
@Arker123 

I think it would be great to review and merge a PR with the color changes *prior* to using rich, as there seems to be wide agreement on that. 

and, let's keep discussing here about:
  1. migrating to `rich` for coloring output, and
  2. more complex drawing that we can do with our console output

once we reach a consensus, then let's open a second PR with the subsequent changes. in essence, i'm excited for your initial work and want to see that get merged right away, and not held up by interim discussions! i also think these are more logical partitions of the changes, so they're more appropriate as separate PRs.

**Comment by williballenthin** [comment]
i've used rich (and [textual](https://textual.textualize.io/)) in another project recently and thought it's a great choice. i like that its actively maintained and handles a lot of the most common configurations very easily. i agree that we should consider replacing our bespoke color management in FLOSS (and capa) with rich.

another benefit: it has a [progress bar widget](https://rich.readthedocs.io/en/stable/progress.html), so we could probably also replace tqdm. and, the hooks into stack traces, logging, etc. might improve the display of other text that occasionally gets rendered.

> One minor concern I have is whether the visual style would hinder some workflows

first, i think we should be wary of introducing too much decoration and work hard to strike a balance between form and function. fortunately i think rich will provide enough flexibility to let us render whatever and however we want.

> we have the -q mode

> Using the file output -o to write

> Creating a different argument

second, i'd point out that rich has built-in [terminal detection](https://rich.readthedocs.io/en/stable/console.html#terminal-detection) which means that it automatically detects if the output is being piped to a file (or pager like less, or grep, etc.) and removes coloring. maybe this is sufficient, which would be great, since FLOSS already has so many CLI arguments. alternatively, perhaps we could use the existing terminal detection to turn on/off additional decoration, such as table headers, progress bars, etc.

thank you @Arker123 for suggesting the idea!

**Comment by williballenthin** [comment]
I think there’s a few things that can be submitted in separate PRs:

  - replace termcolor with rich
  - add rich exception formatter
  - try out rich logging formatter 
  - see if rich can replace tqdm progress bars

There’s also exploring more radical styling changes using new rich functionality. However I’d be wary of going too far down this path without more discussion. Still, it might be fun for you to prototype and share screenshots with the understanding we may not merge all of them.

**Comment by williballenthin** [comment]
@Arker123 an idea: could we use the `rich.stylize()` API instead of using the markup API?
https://rich.readthedocs.io/en/stable/text.html#rich-text

@mr-tz has a good point though: maybe we could use the rich pretty tables with colors instead of mixing with the tabulate library?

**Comment by EmperialX** [comment]
can i work on this issue

**Comment by williballenthin** [comment]
No, @Arker123 is currently assigned to this issue so please give them a chance to implement the feature and close out the issue.

**Comment by EmperialX** [comment]
@williballenthin  Thank you for letting me know, I understand @Arker123 is currently working on this issue and I respect that. Please let me know if there are any other issues or tasks that I can work on to contribute to the project Floss. I look forward to future opportunities. Thank you.

**Comment by mr-tz** [comment]
@Arker123, can this be closed as completed?
